patterns/about.md

```diff
@@ -0,0 +1,137 @@
+	- Restricts accessor/mutator access
+- Proxy
+	- An object representing another object
+
+## Behavioral design patterns
+
+- Chain of responsibility
+	- A way of passing a request between a chain of objects
+- Command
+	- Encapsulate a command request as an object
+- Interpreter
+	- A way to include language elements in a program
+- Iterator
+	- Sequentially access the elements of a collection
+- Mediator
+	- Defines simplified communication between classes
+- Memento
+	- Capture and restore an object's internal state
+- Null Object
+	- Designed to act as a default value of an object
+- Observer
+	- A way of notifying change to a number of classes
+- State
+	- Alter an object's behavior when its state changes
+- Strategy
+	- Encapsulates and algorithm inside a class
+- Template Method
+	- Defer the exact steps of an algorithm to a subclass
+- Visitor
+	- Defines a new operation to a class without change
+
+## Criticism
+
+- Targets the wrong problem
+- Lacks formal foundations
+- Leads to inefficient solutions
+- Does not differ significantly from other abstractions
+
+# Design patterns in Rust
+
+Rust has many very unique features. These features give us great benefit by removing whole classes of problems.
+
+Rust is an imperative language, but it follows many functional programming paradigms. One of the biggest hurdles to understanding functional programs when coming from an imperative background is the shift in thinking. Imperative programs describe __how__ to do something, whereas declarative programs describe __what__ to do. Let's sum the numbers from 1 to 10 to show this.
+
+## Imperative
+
+```rust
+let mut sum = 0;
+for i in 1..11 {
+	sum += i;
+}
+println!("{}", sum);
+```
+
+With imperative programs, we have to play compiler to see what is happening. Here, we start with a `sum` of `0`. Next, we iterate through the range from 1 to 10. Each time through the loop, we add the corresponding value in the range. Then we print it out.
+
+| `i` | `sum` |
+| --- | --- |
+| 1 | 1 |
+| 2 | 3 |
+| 3 | 6 |
+| 4 | 10 |
+| 5 | 15 |
+| 6 | 21 |
+| 7 | 28 |
+| 8 | 36 |
+| 9 | 45 |
+| 10 | 55 |
+
+This is how most of us start out programming. We learn that a program is a set of steps.
+
+## Declarative
+
+```rust
+println!("{}", (1..11).fold(0, |a, b| a + b));
+```
+
+Whoa! This is really different! What's going on here? Remember that with declarative programs we are describing __what__ to do, rather than __how__ to do it. `fold` is a function that [composes](https://en.wikipedia.org/wiki/Function_composition) functions. The name is a convention from Haskell.
+
+Here, we are composing functions of addition (this closure: `|a, b| a + b)`) with a range from 1 to 10. The `0` is the starting point, so `a` is `0` at first. `b` is the first element of the range, `1`. `0 + 1 = 1` is the result. So now we `fold` again, with `a = 1`, `b = 2` and so `1 + 2 = 3` is the next result. This process continues until we get to the last element in the range, `10`.
```

Comment:
Please note that "composing functions" is not really as easy in Rust as in many other functional languages.